### PR TITLE
[SPARK-35558] Optimizes for multi-quantile retrieval

### PR DESCRIPTION
**Pull request by alkispoly-db**
### What changes were proposed in this pull request?
Optimizes the retrieval of approximate quantiles for an array of percentiles.
* Adds an overload for QuantileSummaries.query that accepts an array of percentiles and optimizes the computation to do a single pass over the sketch and avoid redundant computation.
* Modifies the ApproximatePercentiles operator to call into the new method.

All formatting changes are the result of running ./dev/scalafmt

### Why are the changes needed?
The existing implementation does repeated calls per input percentile resulting in redundant computation.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit tests for the new method.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala [hunk 1]
```diff
@@ -229,46 +229,99 @@ class QuantileSummaries(
...
+    val result = Array.fill(percentiles.length)(0.0)
+    sortedPercentiles.foreach {
+      case (percentile, pos) =>
+        if (percentile <= relativeError) {
+          result(pos) = sampled.head.value
+        } else if (percentile >= 1 - relativeError) {
+          result(pos) = sampled.last.value
+        } else {
+          val (newIndex, newMinRank, approxQuantile) =
+            findApproxQuantile(index, minRank, targetError, percentile)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala:301]: I don't need a benchmark or anything, but is this much faster if it calls this method repeatedly? I think it saves some common computation, from what I can see

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by srowen** [comment]
Not sure if it's definitely related, but it looks like this results in tests that hang forever:
`[info] *** Test still running after 16 minutes, 2 seconds: suite name: AdaptiveQueryExecSuite, test name: SPARK-33933: Materialize BroadcastQueryStage first in AQE. `

Not 100% sure how it's connected, but, doesn't seem to be happening on other PRs?

**Comment by srowen** [comment]
Could be related to https://github.com/apache/spark/pull/32725

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala (hunk 1 above)
Comment (alkispoly-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala:301]: If by this method you mean `QuantileSummaries.query` then there is evidence from profiles that this method becomes a bottleneck as the percentile list grows, and in particular the redundant computation seems to be the root cause.

**Comment by alkispoly-db** [comment]
> Could be related to #32725

I can wait until that PR is closed and retest.

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by srowen** [comment]
Merged to master.

**Pull request closed without merging**